### ISSUE TITLE
Add type stubs

### ISSUE THREAD

**Issue by ruancomelli** [Suor/funcy#106]
[Stub files](https://mypy.readthedocs.io/en/stable/stubs.html#stub-files) can be added to the project to provide static type information that can be used by tools like [mypy](https://mypy.readthedocs.io/en/stable/).

I believe that the simplest way to add this functionality to funcy is to provide stub files along with their respective modules. That is, for example, a stub file `colls.pyi` corresponding to the contents of `colls.py`; a `funcs.pyi` corresponding to `funcs.py` and so on. Corresponding snippets from `funcs.py` and `funcs.pyi` would look like:

#### **funcs.py**
```py
def identity(x):
    return x

def constantly(x):
    return lambda *a, **kw: x

def caller(*a, **kw):
    return lambda f: f(*a, **kw)
```
#### **funcs.pyi**
```py
from typing import Any, Callable, TypeVar

T = TypeVar('T')

def identity(x: T) -> T: ...
def constantly(x: T) -> Callable[..., T]: ...
def caller(*a: Any, **kw: Any) -> Callable[[Callable[..., T]], T]: ...
```

Take a look at [more-itertools](https://github.com/more-itertools/more-itertools/tree/master/more_itertools) to see how this looks like when fully implemented.

If you are interested in this functionality, I could perhaps submit a draft PR by the end of the week.

**Comment by ruancomelli** [comment]
Besides code completion, as mentioned by @cardoso-neto, type hints help to ensure code correctness even before execution. Moreover, when programming in typed contexts, funcy does not integrate well and often demands some kind of reworking: I either type-annotate everything manually or ask mypy to skip certain lines or (what I am getting more inclined to do) write stubs for myself.

However, I'm pretty sure that you are already aware of the benefits of adding type annotations everywhere, and the cons of not having them. Now, what are the annoyances of adding them to funcy? I can think of:

1. **possible compatibility issues**: We have to make sure that we will not break older Python versions. There are many workarounds for this, one of them being including stub files (`.pyi`) together with the regular files (`.py`) for the sake of pleasing type-checkers. Those files are ignored at runtime, so no problems here. This is the approach followed by [more-itertools](https://github.com/more-itertools/more-itertools/tree/master/more_itertools), for instance.
2. **increased workload for further developments**: In addition to implementing the desired functionality, contributors have to think about typing as well. While I agree that it is indeed a bit more work, I believe this is acceptable.
3. **possible mismatch between implementation and stubs**: If a contributor changes the signature of a function in the implementation files, but forgets to update the stub files, there will be a mismatch. However, AFAIK type-checkers can be used here to make sure everything is still fine.
4. **writing all of those stub files will take a lot of time right now**: Yes, I strongly agree with you here. But then we can always work on a separate branch and only add it to `main` once we complete this task.

Is there any other reason why you are not a fan of adding types to funcy?

**Comment by Suor** [comment]
Can't one use unions instead of overload? Should be less verbose. Overloading might still be needed for optional first args.

**Comment by Suor** [comment]
It's not `-> Iterable[X] | Iterable[Y]` it's just `-> Iterable[Y]` for the most part. Only `None` is the issue.

**Comment by ruancomelli** [comment]
> Can't one use unions instead of overload? Should be less verbose.
> It's not -> Iterable[X] | Iterable[Y] it's just -> Iterable[Y] for the most part. Only None is the issue.

@Suor perfect; that is why overloads are necessary here, but for sure things can be less verbose. Taking only `None`s, `Callable`s and `Mapping`s into account, we can simplify @cardoso-neto's implementation a bit:
```py
@overload
def funcy_map(f: None, iter_: Iterable[X]) -> Iterable[X]: ...

@overload
def funcy_map(
    f: Callable[[X], Y] | Mapping[X, Y],
    iter_: Iterable[X]
) -> Iterable[Y]: ...
```
Unfortunately, it can't get much simpler than that because there is no way to annotate the return value as being something like `Iterable[X] if f is None else Iterable[Y]`. Whenever the return type changes depending on the inputs, we have to add a new overload.

I wrote a draft implementation for the basic type stubs for the extended function semantics, you can find it in my [gist repository](https://gist.github.com/ruancomelli/d5226de7c65a3c5d233b766bfc09adde) and add comments/suggestions at will. Note that it is an untested draft written for Python 3.9+, an actual implementation should target older Python versions (and be tested, of course!). The main problem I see here is that there are indeed lots of overloads - and I've only written stubs for two functions! The bright side is that new functions will probably be added as copy-paste-tune from those basic versions. We could also modularize things by creating custom generic types, but I'll leave that as an exercise for the reader :)

> Anyway I would prefer someone else, who actually cares about this, support type annotations. Probably outside of funcy repo, i.e. in typeshed.

That is (almost) exactly what I was thinking. I'm not sure about the best way to publish this outside of funcy - I need to do some research first - but typeshed seems to be the way to go indeed. I am someone else who actually cares about this, and I would gladly write and maintain such a thing if you accept it; I'm just very busy at the moment, but perhaps in the next few weeks something could be done. What are your thoughts regarding this?

**Comment by julianfortune** [comment]
@ruancomelli Have you had a chance to start a project for typehints? Thanks!

**Comment by ruancomelli** [comment]
@julianfortune I have implemented some type-stubs locally and was thinking of eventually uploading them to [typeshed](https://github.com/python/typeshed) once I got a complete set of annotations. Unfortunately, I've been in a lack of free time in the past months, so I didn't manage to make this ready for publishing.

Funnily enough, I happened to tweak those annotations yesterday after all of those months. I'm willing to get back to this project in the next few days, but don't refrain from doing it yourself if you wish :grin:

**Comment by julianfortune** [comment]
@ruancomelli I don't think it makes sense to duplicate work, but if you want to make a repo with your work in progress, I would be happy to help out!

**Comment by ruancomelli** [comment]
Hey @julianfortune! It seems that I will not have enough time for this in the next few months, so please feel free to take on this project :grin:

**Comment by Ayenem** [comment]
@julianfortune @ruancomelli Any of you can point me to the repo of the WIP on this? I could contribute here and there.

**Comment by ruancomelli** [comment]
Hi @Ayenem! After your message, I finally took the time to clean up my draft type-annotations and assemble them in a repository: https://github.com/ruancomelli/funcy-stubs.

Feel free to contribute, I would be glad to review and approve PRs!

@julianfortune you may be interested in taking a look at it as well.

Note that a good portion of it is still a draft, unsuitable for release. You can see the modules where I deliberately skipped some functions: I defined a `__getattr__` function in them. In addition, many types can be further narrowed to provide a better type inference for the end users.

Coming soon is a better development setup: I'm planning on writing stubtests and setting up GitHub workflows.

**Comment by Ayenem** [comment]
Perfect  :) Thanks @ruancomelli I'll give it a look

**Comment by bersbersbers** [comment]
> I am not a big fan of adding types to funcy.

I believe they should be added where possible, as adding `funcy` to typed code will break existing type checks. One example:

```python
import numpy as np
from funcy import retry

@retry(3)
def load_data() -> np.typing.NDArray:
    return np.array([0])

# initialize
data: np.typing.NDArray | None = None

# do some work
data = load_data()
np.pad(data, 1)
```

`mypy` says
> Argument 1 to "pad" has incompatible type "Optional[ndarray[Any, dtype[Any]]]"

The example works perfectly when commenting out `@retry`.

I believe typing the decorator aspect of `retry` might be as easy as
```python
from typing import Callable, TypeVar
F = TypeVar("F", bound=Callable)
retry: Callable[..., Callable[[F], F]]
```
(but I may be missing something).

**Comment by bersbersbers** [comment]
> Did you try funcy-stubs linked above?

Only briefly, to be honest. It forces `funcy==1.17`,and even using that (`pip install funcy-stubs`), I still get 

> error: Call to untyped function "fallback" in typed context  [no-untyped-call]

or 

> error: Untyped decorator makes function "load_data" untyped  [misc]

**Comment by bersbersbers** [comment]
A related problem that I have is that I cannot seem to type-annotate the imports myself, see https://github.com/python/mypy/issues/15170

**Comment by ruancomelli** [comment]
Hello, @bersbersbers!
Thanks for the feedback, I'll look into both issues and see if I can solve them in [`funcy-stubs`](https://github.com/ruancomelli/funcy-stubs/issues). Also feel free to open any other typing-related issues in that repository.
Note that the type stubs in `funcy-stubs` is still incomplete, so a few issues are expected. If you have already correctly type-annotated a `funcy` function, please open a pull request in `funcy-stubs` :wink:

**Comment by Suor** [comment]
I have kind of working approach here https://github.com/Suor/funcy/pull/165

**Comment by Suor** [comment]
The actual good approach resulted in so much repetition that I was forced to invent my own language of higher order type annotations - .pyih files, which are transpiled into .pyi. Also a small test framework requiered to keep these in sync and in check, considering all the variations and the mypy, pyright, ty, pyrefly differences.

Started it in January and Opus 4.5 was really struggling and kept making a mess, Opus 5 now, however, finished the thing quite smooth. Given limited time and motivation I would say it won't be feasible before this year and this year AIs, partial attemps both mine and some of others show this.

Anyway, if anyone interested can give a shot using the branch from above.
